### PR TITLE
Improve suggestions for broken intra-doc links

### LINKED ISSUE #75305: Improve diagnostics when intra-doc-resolution fails
Similar to but not the same as https://github.com/rust-lang/rust/issues/74207. Right now rustdoc reports the same error for all the following cases:

```rust
/// [<invalid syntax>]
/// [path::to::nonexistent::module]
/// [f::A]
/// [S::A]
/// [S::fmt]
pub fn f() {}
#[derive(Debug)]
pub struct S;
```
```
warning: unresolved link to `path::to::nonexistent::module`
 --> /dev/stdin:2:6
  |
2 | /// [path::to::nonexistent::module]
  |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ unresolved link
  |
  = note: `#[warn(broken_intra_doc_links)]` on by default
  = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
```

However we can give much more helpful errors:
```
1 | /// [<invalid syntax>]
  |     ^^^^^^^^^^^^^^^^^^ not valid syntax for an intra-doc link
2 | /// [path::to::nonexistent::module]
  |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ use of undeclared type or module `path` 
3 | /// [f::A]
  |     ^^^^^^ functions cannot have associated items
4 | /// [S::A]
  |     ^^^^^^ `S` does not have an associated item `A` in scope. help: try importing a trait that defines that item
5 | /// [S::fmt]
  |     ^^^^^^ `S` does not have an associated item `fmt` in scope. help: S implements `std::fmt::Debug`, try importing it.
```

Most of these do not require special help from `rustc_resolve` (that would just let rustdoc duplicate less code).

### PR DESCRIPTION
**Pull request by jyn514**
~~Depends on #74489 and should not be merged before that PR.~~ Merged 🎉
~~Depends on #75916 and should not be merged before.~~ Merged

Fixes https://github.com/rust-lang/rust/issues/75305.

This does a lot of different things 😆.

- Add `PerNS::into_iter()` so I didn't have to keep rewriting hacks around it. Also add `PerNS::iter()` for consistency. Let me know if this should be `impl IntoIterator` instead.
- Make `ResolutionFailure` an enum instead of a unit variant. This was most of the changes: everywhere that said `ErrorKind::ResolutionFailure` now has to say _why_ the link failed to resolve.
- Store the resolution in case of an anchor failure. Previously this was implemented as variants on `AnchorFailure` which was prone to typos and had inconsistent output compared to the rest of the diagnostics.
- Turn some `Err`ors into unwrap() or panic()s, because they're rustdoc bugs and not user error. These have comments as to why they're bugs (in particular this would have caught #76073 as a bug a while ago).
- If an item is not in scope at all, say the first segment in the path that failed to resolve
- If an item exists but not in the current namespaces, say that and suggests linking to that namespace.
- If there is a partial resolution for an item (part of the segments resolved, but not all of them), say the partial resolution and why the following segment didn't resolve.
- Add the `DefId` of associated items to `kind_side_channel` so it can be used for diagnostics (tl;dr of the hack: the rest of rustdoc expects the id of the item, but for diagnostics we need the associated item).
- No longer suggests escaping the brackets for every link that failed to resolve; this was pretty obnoxious. Now it only suggests `\[ \]` if no segment resolved and there is no `::` in the link.
- Add `Suggestion`, which says _what_ to prefix the link with, not just 'prefix with the item kind'.

Places where this is currently buggy:

<details><summary>All outdated</summary>

~~1. When the link has the wrong namespace:~~ Now fixed.

<details>

```rust
/// [type@S::h]
impl S {
	pub fn h() {}
}

/// [type@T::g]
pub trait T {
	fn g() {}
}
```
```
error: unresolved link to `T::g`
  --> /home/joshua/rustc/src/test/rustdoc-ui/intra-link-errors.rs:53:6
   |
53 | /// [type@T::g]
   |      ^^^^^^^^^
   |
   = note: this link partially resolves to the trait `T`,
   = note: `T` has no field, variant, or associated item named `g`

error: unresolved link to `S::h`
  --> /home/joshua/rustc/src/test/rustdoc-ui/intra-link-errors.rs:48:6
   |
48 | /// [type@S::h]
   |      ^^^^^^^^^
   |
   = note: this link partially resolves to the struct `S`,
   = note: `S` has no field, variant, or associated item named `h`
```
Instead it should suggest changing the disambiguator, the way it currently does for macros:
```
error: unresolved link to `S`
  --> /home/joshua/rustc/src/test/rustdoc-ui/intra-link-errors.rs:38:6
   |
38 | /// [S!]
   |      ^^ help: to link to the unit struct, use its disambiguator: `value@S`
   |
   = note: this link resolves to the unit struct `S`, which is not in the macro namespace
```

</details>

2. ~~Associated items for values. It says that the value isn't in scope; instead it should say that values can't have associated items.~~ Fixed.

<details>

```
error: unresolved link to `f::A`
  --> /home/joshua/rustc/src/test/rustdoc-ui/intra-link-errors.rs:14:6
   |
14 | /// [f::A]
   |      ^^^^
   |
   = note: no item named `f` is in scope
   = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
```
This is _mostly_ fixed, it now says

```rust
warning: unresolved link to `f::A`
 --> /home/joshua/test-rustdoc/f.rs:1:6
  |
1 | /// [f::A]
  |      ^^^^
  |
  = note: this link partially resolves to the function `f`
  = note: `f` is a function, not a module
```

'function, not a module' seems awfully terse when what I actually mean is '`::` isn't allowed here', though.

</details>

It looks a lot nicer now, it says

```
error: unresolved link to `f::A`
  --> /home/joshua/rustc/src/test/rustdoc-ui/intra-link-errors.rs:13:6
   |
13 | /// [f::A]
   |      ^^^^
   |
   = note: `f` is a function, not a module or type, and cannot have associated items
```

3. ~~I'm also not very happy with the second note for this error:~~

<details>
```
error: unresolved link to `S::A`
  --> /home/joshua/rustc/src/test/rustdoc-ui/intra-link-errors.rs:19:6
   |
19 | /// [S::A]
   |      ^^^^
   |
   = note: this link partially resolves to the struct `S`,
   = note: `S` has no field, variant, or associated item named `A`
```

but I'm not sure how better to word it.

I ended up going with 'no `A` in `S`' to match `rustc_resolve` but that seems terse as well.

</details>

This now says

```
error: unresolved link to `S::A`
  --> /home/joshua/rustc/src/test/rustdoc-ui/intra-link-errors.rs:17:6
   |
17 | /// [S::A]
   |      ^^^^
   |
   = note: the struct `S` has no field or associated item named `A`
```

which I think looks pretty good :)

4. This is minor, but it would be nice to say that `path` wasn't found instead of the full thing:
```
error: unresolved link to `path::to::nonexistent::module`
 --> /home/joshua/rustc/src/test/rustdoc-ui/intra-link-errors.rs:8:6
  |
8 | /// [path::to::nonexistent::module]
  |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

It will now look at most 3 paths up (so it reports `path::to` as not in scope), but it doesn't work with arbitrarily many paths.

</details>

~~I recommend only reviewing the last few commits - the first 7 are all from #74489.~~ Rebased so that only the relevant commits are shown. Let me know if I should squash the history some more.

r? @estebank

### CODE REVIEW

File: src/test/rustdoc-ui/intra-link-errors.stderr [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+   = note: `f` is a function, not a module or type, and cannot have associated items
+
+error: unresolved link to `S::A`
+  --> $DIR/intra-link-errors.rs:19:6
+   |
+LL | /// [S::A]
+   |      ^^^^
+   |
+   = note: this link partially resolves to the struct `S`
+   = note: no `A` in `S`
```
Comment (jyn514) [inline src/test/rustdoc-ui/intra-link-errors.stderr:31]: A rustacean on discord suggested that this be consistent with the more verbose form for primitives:
```suggestion
error: unresolved link to `S::A`
  --> $DIR/intra-link-errors.rs:19:6
   |
LL | /// [S::A]
   |      ^^^^
   |
   = note: this link partially resolves to the struct `S`
   = note: `S` does not have a field or associated item named `fmt`
```

**Comment by jyn514** [comment]
@estebank This is ready for review.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 2]
```diff
@@ -131,18 +191,46 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
+                    &intermediate_path,
+                    Some(module_id),
+                    current_item,
+                    extra_fragment,
+                ) {
+                    ResolutionFailure::NoAssocItem(intermediate, variant_field_name)
+                } else {
+                    // Even with the shorter path, it didn't resolve, so say that.
+                    ResolutionFailure::NoAssocItem(ty_res, variant_name)
+                };
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:222]: Actually this would be a good place to look up the first path segment that didn't resolve - just make this a loop instead of doing it once:
```rust
let err = loop {
     if let Some(res) = self.check_full_res(intermediate_path) {
        break res;
    }
    intermediate_path = match intermediate_path.rsplitn(2, "::").nth(1) {
       None => break ResolutionFailure::NoAssocItem(ty_res, variant_name),
       Some(inner) => inner,
    };
}
```

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 3]
```diff
@@ -270,21 +371,52 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
+                Err(()) | Ok(Res::Err) => {
+                    return if ns == Namespace::ValueNS {
+                        self.variant_field(path_str, current_item, module_id, extra_fragment)
+                    } else {
+                        // See if it only broke because of the namespace.
+                        let kind = cx.enter_resolver(|resolver| {
+                            for &ns in &[MacroNS, ValueNS] {
+                                match resolver
+                                    .resolve_str_path_error(DUMMY_SP, &path_root, ns, module_id)
+                                {
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:403]: This should probably use `check_full_res` instead.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 4]
```diff
@@ -1241,10 +1580,10 @@ fn anchor_failure(
     });
 }
 
-fn ambiguity_error(
+fn ambiguity_error<'a>(
     cx: &DocContext<'_>,
     item: &Item,
-    path_str: &str,
+    path_str: &'a str,
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:1586]: This can get rid of the lifetime parameter now that it returns `ErrorKind<'static>`.

**Review by jyn514 — COMMENTED** [review]
Left myself some comments 😆

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 3 above)
Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs:413]: Actually it shouldn't - I left a comment why.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 2 above)
Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs:231]: It shouldn't go here, but slightly above where `ty_res` failed to resolve. I implemented this :)

**Comment by jyn514** [comment]
So I don't lose it: @pickfire suggested improving the description of how to add a disambiguator: https://github.com/rust-lang/rust/pull/75916#discussion_r477177431

> Would be be better to be clearer to put the real item kind rather than just the text "item kind"?
```
 help: to link to the enum, prefix with 'enum@'
```
Currently it says
```
 help: to link to the enum, prefix with the item kind
```
which is a little less specific.

File: src/test/rustdoc-ui/intra-link-errors.stderr [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+LL | /// [path::to::nonexistent::module]
+   |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+note: the lint level is defined here
+  --> $DIR/intra-link-errors.rs:1:9
+   |
+LL | #![deny(broken_intra_doc_links)]
+   |         ^^^^^^^^^^^^^^^^^^^^^^
+   = note: no item named `path::to` is in scope
+   = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/test/rustdoc-ui/intra-link-errors.stderr]: Implemented, I think the benefits outweigh the edge cases.

**Comment by jyn514** [comment]
> Would be be better to be clearer to put the real item kind rather than just the text "item kind"?


Implemented this :)

File: src/test/rustdoc-ui/intra-link-errors.stderr [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+   = note: this link resolves to the associated function `g`, which is not in the type namespace
+
+error: unresolved link to `T::h`
+  --> $DIR/intra-link-errors.rs:68:6
+   |
+LL | /// [T::h!]
+   |      ^^^^^
+   |
+   = note: no item named `T::h` is in scope
+   = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/test/rustdoc-ui/intra-link-errors.stderr]: Fixed! And there are now no more review comments left :)

```
error: unresolved link to `T::h`
  --> $DIR/intra-link-errors.rs:67:6
   |
LL | /// [T::h!]
   |      ^^^^^
   |
   = note: the trait `T` has no macro named `h`
```

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 7]
```diff
@@ -99,14 +153,38 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
+        let ty_res = cx
             .enter_resolver(|resolver| {
                 resolver.resolve_str_path_error(DUMMY_SP, &path, TypeNS, module_id)
             })
-            .map_err(|_| ErrorKind::ResolutionFailure)?;
+            .map(|(_, res)| res)
+            .unwrap_or(Res::Err);
+        // This code only gets hit if three path segments in a row don't get resolved.
+        // It's a good time to check if _any_ parent of the path gets resolved.
+        // If so, report it and say the first which failed; if not, say the first path segment didn't resolve.
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:167]: Hmm, I just realized this will only run in the value namespace, so if you have a type or macro disambiguator it will still not show the first segment that failed to resolve. I should probably fix that.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 8]
```diff
@@ -118,7 +196,9 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
                     .flat_map(|imp| cx.tcx.associated_items(*imp).in_definition_order())
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:196]: This is buggy: it should use `find_by_name_and_namespace` to avoid false positives.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 9]
```diff
@@ -118,7 +196,9 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
                     .any(|item| item.ident.name == variant_name)
                 {
-                    return Err(ErrorKind::ResolutionFailure);
+                    // This is just to let `fold_item` know that this shouldn't be considered;
+                    // it's a bug for the error to make it to the user
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:200]:
```suggestion
                    // This was already found in another namespace. No need to return it again.
                    // This error is just to let `fold_item` know that this shouldn't be considered;
                    // it's a bug for the error to make it to the user
```
Hmm, on reflection it seems like there is some ambiguity here. Consider ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=24e1246fd29017cc498b98a6a2c80841)):
```rust
#[derive(Debug)]
#[allow(non_camel_case_types)]
pub enum E { f, g }
impl E { pub fn f() {} }

fn main() {
    println!("{:?}", E::f);
    println!("{:?}", E::f()); // error[E0618]: expected function, found enum variant `E::f`
    let f: fn() = E::f;  // error[E0308]: mismatched types: expected fn pointer, found enum `E`
    println!("{:?}", f);
}
```
`E::f` (the variant) and `E::f()` (the function) are both in the same namespace. It looks like the way `rustc` treats this is to just never let you access `E::f()`?? This is weird.

Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs:200]: To be consistent with rustc this should probably be switched in rustdoc so that `E::f` (the variant) has precedence. But this seems like a really weird part of the language.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 10]
```diff
@@ -131,18 +211,46 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
                         } else {
-                            Err(ErrorKind::ResolutionFailure)
+                            Err(ErrorKind::Resolve(ResolutionFailure::NotAVariant(
+                                ty_res,
+                                variant_field_name,
+                            )))
                         }
                     }
-                    _ => Err(ErrorKind::ResolutionFailure),
+                    _ => unreachable!(),
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:220]:
```suggestion
                    _ => unreachable!("type_of(Res::Def(Enum, _)) should always be an enum"),
```

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 11]
```diff
@@ -168,25 +276,44 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
                     if let Res::Def(..) = res {
                         let res = res.map_id(|_| panic!("unexpected node_id"));
-                        return Some(res);
+                        return Some(Ok(res));
                     }
                 }
             } else {
                 debug!("attempting to resolve item without parent module: {}", path_str);
+                return Some(Err(ResolutionFailure::NoParentItem));
             }
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:285]: At some point I should probably refactor this out so it only happens in one place. The only time this should happen is for synthetic impls or something which can't be documented anyway.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 12]
```diff
@@ -168,25 +276,44 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
             None
         })
+        // This weird control flow is so we don't borrow the resolver more than once at a time
+        .unwrap_or_else(|| {
+            let mut split = path_str.rsplitn(2, "::");
+            if let Some((parent, base)) = split.next().and_then(|x| Some((split.next()?, x))) {
+                if let Some(res) = self.check_full_res(TypeNS, parent, parent_id, &None, &None) {
+                    return Err(if matches!(res, Res::PrimTy(_)) {
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:293]:
```suggestion
                    // Check for `PrimTy` so we can distinguish `NoPrimitiveAssocItem`.
                    // `resolution_failure` needs this so it can give the name of the primitive;
                    // `prim.descr()` just says "builtin type".
                    return Err(if matches!(res, Res::PrimTy(_)) {
```

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 13]
```diff
@@ -168,25 +276,44 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
+                        return Some(Ok(res));
                     }
                 }
             } else {
                 debug!("attempting to resolve item without parent module: {}", path_str);
+                return Some(Err(ResolutionFailure::NoParentItem));
             }
             None
         })
+        // This weird control flow is so we don't borrow the resolver more than once at a time
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:288]:
```suggestion
        // Look for partial resolutions if this failed to resolve.
        // This weird control flow is so we don't borrow the resolver more than once at a time
```

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 14]
```diff
@@ -272,21 +411,53 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
-            }
+                .map(|(_, res)| res);
+            let ty_res = match ty_res {
+                Err(()) | Ok(Res::Err) => {
+                    return if ns == Namespace::ValueNS {
+                        self.variant_field(path_str, current_item, module_id, extra_fragment)
+                    } else {
+                        // See if it only broke because of the namespace.
+                        let kind = cx.enter_resolver(|resolver| {
+                            // NOTE: this doesn't use `check_full_res` because we explicitly want to ignore `TypeNS` (we already checked it)
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:440]:
```suggestion
                            // NOTE: this doesn't use `check_full_res` because we explicitly want to ignore `TypeNS` (we already checked it)
                            // In particular, checking `TypeNS` could cause infinite recursion, because `TypeNS` ends up back here.
```

**Review by jyn514 — COMMENTED** [review]
~~lol I don't know how to use github~~

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 8 above)
Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs:196]: I think fixing this should go in a follow-up PR.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 9 above)
Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs]: I think fixing this should go in a follow-up PR, it's a behavior change and not really related to diagnostics.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 11 above)
Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs:286]: I think fixing this should go in a follow-up PR - it's a easy change but it will be a lot of churn, making this harder to review.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 7 above)
Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs]: Implemented in https://github.com/rust-lang/rust/pull/75756/commits/c55f47b2081a1610559ed0ce0f028c8fb678eab1.

**Comment by jyn514** [comment]
Ok, this is _really_ ready for review this time 😆

**Comment by jyn514** [comment]
ping @estebank, it's been a week or so. I know this is a bit of a giant PR - would it be easier to only review the diagnostics themselves and I can get a rustdoc member to review the logic changes?

**Comment by estebank** [comment]
Sorry, reviewing.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 15]
```diff
@@ -807,48 +1037,62 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
...
+                            );
                             // this could just be a normal link
                             continue;
                         }
 
-                        let len = candidates.clone().present_items().count();
-
                         if len == 1 {
-                            candidates.present_items().next().unwrap()
+                            candidates_iter.next().unwrap().clone()
```
Comment (estebank) [inline src/librustdoc/passes/collect_intra_doc_links.rs:1084]: Why is the `clone` needed?

This is the happy path, right?

File: src/test/rustdoc-ui/assoc-item-not-in-scope.stderr [hunk 16]
```diff
@@ -2,14 +2,14 @@ error: unresolved link to `S::fmt`
...
-   |              ^^^^^^^^ unresolved link
+   |              ^^^^^^^^
    |
 note: the lint level is defined here
   --> $DIR/assoc-item-not-in-scope.rs:1:9
    |
 LL | #![deny(broken_intra_doc_links)]
    |         ^^^^^^^^^^^^^^^^^^^^^^
-   = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
+   = note: the struct `S` has no field or associated item named `fmt`
```
Comment (estebank) [inline src/test/rustdoc-ui/assoc-item-not-in-scope.stderr:12]: Let's make this a `span_label` on the main span.

File: src/test/rustdoc-ui/deny-intra-link-resolution-failure.stderr [hunk 17]
```diff
@@ -2,13 +2,14 @@ error: unresolved link to `v2`
...
 LL | /// [v2]
-   |      ^^ unresolved link
+   |      ^^
    |
 note: the lint level is defined here
   --> $DIR/deny-intra-link-resolution-failure.rs:1:9
    |
 LL | #![deny(broken_intra_doc_links)]
    |         ^^^^^^^^^^^^^^^^^^^^^^
+   = note: no item named `v2` is in scope
```
Comment (estebank) [inline src/test/rustdoc-ui/deny-intra-link-resolution-failure.stderr:12]: Make this a `span_label` on the main span. We probably want to provide some extra context about what the current scope is, but that can be done in a separate PR.

File: src/test/rustdoc-ui/intra-doc-alias-ice.stderr [hunk 18]
```diff
@@ -2,14 +2,14 @@ error: unresolved link to `TypeAlias::hoge`
...
-   |                              ^^^^^^^^^^^^^^^ unresolved link
+   |                              ^^^^^^^^^^^^^^^
    |
 note: the lint level is defined here
   --> $DIR/intra-doc-alias-ice.rs:1:9
    |
 LL | #![deny(broken_intra_doc_links)]
    |         ^^^^^^^^^^^^^^^^^^^^^^
-   = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
+   = note: the type alias `TypeAlias` has no associated item named `hoge`
```
Comment (estebank) [inline src/test/rustdoc-ui/intra-doc-alias-ice.stderr:12]: Make this a `span_label` on the main span.

File: src/test/rustdoc-ui/intra-link-errors.stderr [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+   |
+LL | /// [path::to::nonexistent::module]
+   |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+note: the lint level is defined here
+  --> $DIR/intra-link-errors.rs:1:9
+   |
+LL | #![deny(broken_intra_doc_links)]
+   |         ^^^^^^^^^^^^^^^^^^^^^^
+   = note: no item named `path` is in scope
```
Comment (estebank) [inline src/test/rustdoc-ui/intra-link-errors.stderr:12]: Make this a `span_label` on the main span.

**Review by estebank — COMMENTED** [review]
When it comes to the diagnostic output, the main change I'd make is turn the new `note`s into the main span_label. Ideally, we would create a new span for cases like `[foo::bar]` to point _only_ at the part of the path that is problematic, be it `foo` or `bar`, but that can be addressed later.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 20]
```diff
@@ -78,17 +127,25 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
         let cx = self.cx;
 
+        debug!("looking for enum variant {}", path_str);
         let mut split = path_str.rsplitn(3, "::");
-        let variant_field_name =
-            split.next().map(|f| Symbol::intern(f)).ok_or(ErrorKind::ResolutionFailure)?;
+        let variant_field_name = split
+            .next()
+            .map(|f| Symbol::intern(f))
+            .expect("fold_item should ensure link is non-empty");
```
[the remark this answers is not in the document]
Comment (estebank) [reply src/librustdoc/passes/collect_intra_doc_links.rs:142]: Based on the description, resolving...

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 15 above)
Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs]: This is the happy path, yes. It's not, I wasn't thinking very clearly while writing this. I change this to remove the `clone()` and also the confusing `drop(candidates_iter)` in the other paths.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 21]
```diff
@@ -46,19 +47,67 @@ pub fn collect_intra_doc_links(krate: Crate, cx: &DocContext<'_>) -> Crate {
     }
 }
 
-enum ErrorKind {
-    ResolutionFailure,
+enum ErrorKind<'a> {
+    Resolve(ResolutionFailure<'a>),
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs]: Done, good idea. That said, intra_doc_links are pretty non-performant ... but it doesn't help to make it even worse.

File: src/test/rustdoc-ui/deny-intra-link-resolution-failure.stderr (hunk 17 above)
Comment (jyn514) [reply src/test/rustdoc-ui/deny-intra-link-resolution-failure.stderr]: > We probably want to provide some extra context about what the current scope is, but that can be done in a separate PR

This ended up being really easy to do so I added it here. Let me know if I should separate it out.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs:148]: > The previous behavior might have been masking incorrect rustc behavior, but it would now be an ICE

Right, and that's intentional because it's a bug. There's no reasonable error we could give here other than 'rustdoc has buggy logic', which I think is better served by an ICE.

**Comment by jyn514** [comment]
Thanks so much for the review!
